kong/db/dao/plugins/external.lua

```diff
@@ -0,0 +1,642 @@
+  if type(infos) ~= "table" then
+    error(string.format("Not a plugin info table: \n%s\n%s",
+        server_def.query_command, infos_dump))
+    return
+  end
+
+  for _, plugin_info in ipairs(infos) do
+    register_plugin_info(server_def, plugin_info)
+  end
+end
+
+local function load_all_infos()
+  if not _plugin_infos then
+    _plugin_infos = {}
+
+    for _, server_def in ipairs(get_server_defs()) do
+      ask_info(server_def)
+    end
+  end
+
+  return _plugin_infos
+end
+
+
+local loaded_plugins = {}
+
+local function get_plugin(plugin_name)
+  if not loaded_plugins[plugin_name] then
+    local plugin = load_all_infos()[plugin_name]
+    loaded_plugins[plugin_name] = build_phases(plugin)
+  end
+
+  return loaded_plugins[plugin_name]
+end
+
+function external_plugins.load_plugin(plugin_name)
+  local plugin = get_plugin(plugin_name)
+  if plugin and plugin.PRIORITY then
+    return true, plugin
+  end
+
+  return false, "no plugin found"
+end
+
+function external_plugins.load_schema(plugin_name)
+  local plugin = get_plugin(plugin_name)
+  if plugin and plugin.PRIORITY then
+    return true, plugin.schema
+  end
+
+  return false, "no plugin found"
+end
+
+
+--[[
+
+Process management
+
+Servers that specify an `.exec` field are launched and managed by Kong.
+This is an attempt to duplicate the smallest reasonable subset of systemd.
+
+Each process specifies executable, arguments and environment.
+Stdout and stderr are joined and logged, if it dies, Kong logs the event
+and respawns the server.
+
+--]]
+
+local function grab_logs(proc, name)
+  local prefix = string.format("[%s:%d] ", name, proc:pid())
+
+  while true do
+    local data, err, partial = proc:stdout_read_line()
+    local line = data or partial
+    if line and line ~= "" then
+      raw_log(ngx_INFO, prefix .. line)
+    end
+
+    if not data and err == "closed" then
+      return
+    end
+  end
+end
+
+local function handle_server(server_def)
+  if not server_def.socket then
+    -- no error, just ignore
+    return
+  end
+
+  if server_def.start_command then
+    ngx_timer_at(0, function(premature)
+      if premature then
+        return
+      end
+
+      local ngx_pipe = require "ngx.pipe"
+
+      while not ngx.worker.exiting() do
+        kong.log.notice("Starting " .. server_def.name or "")
+        server_def.proc = assert(ngx_pipe.spawn(server_def.start_command, {
+          merge_stderr = true,
+        }))
+        server_def.proc:set_timeouts(nil, nil, nil, 0)     -- block until something actually happens
+
+        while true do
+          grab_logs(server_def.proc, server_def.name)
+          local ok, reason, status = server_def.proc:wait()
+          if ok ~= nil or reason == "exited" then
+            kong.log.notice("external pluginserver '", server_def.name, "' terminated: ", tostring(reason), " ", tostring(status))
+            break
+          end
+        end
+      end
+      kong.log.notice("Exiting: go-pluginserver not respawned.")
```

Comment:
Probably a copy-paste typo:

```suggestion
      kong.log.notice("Exiting: plugin server not respawned.")
```